internal/internal_eager_workflow.go

```diff
@@ -0,0 +1,94 @@
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"math/rand"
+	"sync"
+	"sync/atomic"
+
+	"go.temporal.io/api/workflowservice/v1"
+)
+
+// eagerWorkflowDispatcher is responsible for finding an available worker for an eager workflow task.
+type eagerWorkflowDispatcher struct {
+	lock               sync.Mutex
+	workersByTaskQueue map[string][]eagerWorker
+}
+
+// registerWorker registers a worker that can be used for eager workflow dispatch
+func (e *eagerWorkflowDispatcher) registerWorker(worker *workflowWorker) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	e.workersByTaskQueue[worker.executionParameters.TaskQueue] = append(e.workersByTaskQueue[worker.executionParameters.TaskQueue], worker.worker)
+}
+
+// applyToRequest updates request if eager workflow dispatch is possible and returns the eagerWorkflowExecutor to use
+func (e *eagerWorkflowDispatcher) applyToRequest(request *workflowservice.StartWorkflowExecutionRequest) *eagerWorkflowExecutor {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	// Try every worker that is assigned to the desired task queue.
+	workers := e.workersByTaskQueue[request.GetTaskQueue().Name]
+	rand.Shuffle(len(workers), func(i, j int) { workers[i], workers[j] = workers[j], workers[i] })
+	for _, worker := range workers {
+		if worker.tryReserveSlot() {
+			request.RequestEagerExecution = true
+			return &eagerWorkflowExecutor{
+				worker: worker,
+			}
+		}
+	}
+	return nil
+}
+
+// eagerWorkflowExecutor is a worker-scoped executor for an eager workflow task.
+type eagerWorkflowExecutor struct {
+	handledResponse atomic.Bool
+	worker          eagerWorker
+}
+
+// handleResponse of an eager workflow task from a StartWorkflowExecution request.
+func (e *eagerWorkflowExecutor) handleResponse(response *workflowservice.PollWorkflowTaskQueueResponse) {
+	if !e.handledResponse.CompareAndSwap(false, true) {
+		panic("eagerWorkflowExecutor trying to handle multiple responses")
+	}
```

Comment:
What is the situation you are trying to protect from here?

Reply:
a caller accidentally reusing  eagerWorkflowExecutor

Reply:
I think we should be allowed to trust callers on non-exported things, i.e. ourselves, here

Reply:
I don't